src/widgets/widget_list/widget_list_item.rs

```diff
@@ -0,0 +1,31 @@
+use crate::{
+    buffer::Buffer,
+    layout::{Rect, Size},
+    widgets::{BarChart, Gauge, Paragraph, SizeHint, Widget},
+};
+
+pub enum WidgetListItem<'a> {
```

Comment:
you are right. the `WidgetListItem` is more an example for the user how you would solve this. maybe I should move it from the library to the examples?

I was also playing with the `Box<dyn Widget + SizeHint>` idea. But then I didn't want to allocate memory every time when rendering. But  maybe it is fine. I will try to play more with this idea

Reply:
> I was playing with this idea a bit last night, and it's a tricky proposition. Something like `Box<dyn Widget + SizeHint>` runs into ownership issues calling render. I have a couple other ideas to try out (e.g. a closure based approach might be promising) - I'll let you know if I think of anything that works, but I wanted to call it out here in case someone else has ideas on that too.

I also managed to get that approach working, but instead of with a new type, using the standard `List` while still being mostly backwards compatible* with existing usages (the trick is to change the code that accepts `Into<Text>` to be `Into<SomeWrapperType>` where `SomeWrapperType` just implements all the same `From<T>`s as `Text` does and you end up with . But I also ran into that same `render(self, ...)` issue you mentioned that prevents this on generic Widgets.

```diff
diff --git a/src/widgets/list.rs b/src/widgets/list.rs
index d6e58c7..adfdbe7 100644
--- a/src/widgets/list.rs
+++ b/src/widgets/list.rs
@@ -2,11 +2,13 @@ use crate::{
     buffer::Buffer,
     layout::{Corner, Rect},
     style::Style,
-    text::Text,
+    text::{Spans, Text},
     widgets::{Block, StatefulWidget, Widget},
 };
 use unicode_width::UnicodeWidthStr;
 
+use super::Paragraph;
+
 #[derive(Debug, Clone, Default)]
 pub struct ListState {
     offset: usize,
@@ -26,16 +28,97 @@ impl ListState {
     }
 }
 
-#[derive(Debug, Clone, PartialEq, Eq)]
 pub struct ListItem<'a> {
-    content: Text<'a>,
+    content: ListItemWidgetWrapper<'a>,
     style: Style,
 }
 
+#[non_exhaustive]
+pub enum ListItemWidgetWrapper<'a> {
+    Text(TextListItemWidget<'a>),
+    Paragraph(Paragraph<'a>),
+    Widget(Box<dyn Widget + 'a>),
+}
+
+impl<'a> From<Vec<Spans<'a>>> for ListItemWidgetWrapper<'a> {
+    fn from(value: Vec<Spans<'a>>) -> Self {
+        Self::Text(TextListItemWidget::new(value))
+    }
+}
+
+impl<'a> From<Spans<'a>> for ListItemWidgetWrapper<'a> {
+    fn from(value: Spans<'a>) -> Self {
+        Self::Text(TextListItemWidget::new(value))
+    }
+}
+
+impl<'a> From<&'a str> for ListItemWidgetWrapper<'a> {
+    fn from(value: &'a str) -> Self {
+        Self::Text(TextListItemWidget::new(value))
+    }
+}
+
+impl<'a> From<Paragraph<'a>> for ListItemWidgetWrapper<'a> {
+    fn from(value: Paragraph<'a>) -> Self {
+        Self::Paragraph(value)
+    }
+}
+
+impl<'a> ListItemWidgetWrapper<'a> {
+    fn height(&self) -> usize {
+        match self {
+            // note I made Paragraph.text public for demo purposes
+            // The actual impl would let paragraph calculate height
+            // and render itself properly
+            ListItemWidgetWrapper::Text(t) => t.0.height(),
+            ListItemWidgetWrapper::Paragraph(p) => p.text.height(),
+            ListItemWidgetWrapper::Widget(_) => todo!(),
+        }
+    }
+
+    fn width(&self) -> usize {
+        match self {
+            ListItemWidgetWrapper::Text(t) => t.0.width(),
+            ListItemWidgetWrapper::Paragraph(p) => p.text.width(),
+            ListItemWidgetWrapper::Widget(_) => todo!(),
+        }
+    }
+
+    fn lines(&self) -> Vec<Spans<'a>> {
+        match self {
+            ListItemWidgetWrapper::Text(t) => t.0.lines.clone(),
+            ListItemWidgetWrapper::Paragraph(p) => p.text.lines.clone(),
+            ListItemWidgetWrapper::Widget(_) => todo!(),
+        }
+    }
+
+    fn render(&self, area: Rect, buffer: &mut Buffer) {
+        match self {
+            ListItemWidgetWrapper::Text(t) => t.clone().render(area, buffer),
+            ListItemWidgetWrapper::Paragraph(p) => p.clone().render(area, buffer),
+            ListItemWidgetWrapper::Widget(_) => todo!(),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct TextListItemWidget<'a>(Text<'a>);
+impl<'a> TextListItemWidget<'a> {
+    fn new(value: impl Into<Text<'a>>) -> Self {
+        Self(value.into())
+    }
+}
+
+impl Widget for TextListItemWidget<'_> {
+    fn render(self, area: Rect, buf: &mut Buffer) {
+        Paragraph::new(self.0).render(area, buf);
+    }
+}
+
 impl<'a> ListItem<'a> {
     pub fn new<T>(content: T) -> ListItem<'a>
     where
-        T: Into<Text<'a>>,
+        T: Into<ListItemWidgetWrapper<'a>>,
     {
         ListItem {
             content: content.into(),
@@ -55,6 +138,10 @@ impl<'a> ListItem<'a> {
     pub fn width(&self) -> usize {
         self.content.width()
     }
+
+    pub fn render(&self, area: Rect, buffer: &mut Buffer) {
+        self.content.render(area, buffer);
+    }
 }
 
 /// A widget to display several items among which one can be selected (optional)
@@ -71,7 +158,6 @@ impl<'a> ListItem<'a> {
 ///     .highlight_style(Style::default().add_modifier(Modifier::ITALIC))
 ///     .highlight_symbol(">>");
 /// ```
-#[derive(Debug, Clone)]
 pub struct List<'a> {
     block: Option<Block<'a>>,
     items: Vec<ListItem<'a>>,
@@ -230,7 +316,7 @@ impl<'a> StatefulWidget for List<'a> {
             buf.set_style(area, item_style);
 
             let is_selected = state.selected.map(|s| s == i).unwrap_or(false);
-            for (j, line) in item.content.lines.iter().enumerate() {
+            for (j, _line) in item.content.lines().iter().enumerate() {
                 // if the item is selected, we need to display the highlight symbol:
                 // - either for the first line of the item only,
                 // - or for each line of the item if the appropriate option is set
@@ -239,7 +325,7 @@ impl<'a> StatefulWidget for List<'a> {
                 } else {
                     &blank_symbol
                 };
-                let (elem_x, max_element_width) = if has_selection {
+                let (_elem_x, _max_element_width) = if has_selection {
                     let (elem_x, _) = buf.set_stringn(
                         x,
                         y + j as u16,
@@ -251,8 +337,18 @@ impl<'a> StatefulWidget for List<'a> {
                 } else {
                     (x, list_area.width)
                 };
-                buf.set_spans(elem_x, y + j as u16, line, max_element_width);
             }
+            let symbol_width = if has_selection {
+                self.highlight_symbol.map_or(0, |s| s.width()) as u16
+            } else {
+                0
+            };
+            let item_area = Rect {
+                x: area.x + symbol_width,
+                width: area.width - symbol_width,
+                ..area
+            };
+            item.render(item_area, buf);
             if is_selected {
                 buf.set_style(area, self.highlight_style);
             }
@@ -266,3 +362,39 @@ impl<'a> Widget for List<'a> {
         StatefulWidget::render(self, area, buf, &mut state);
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::assert_buffer_eq;
+
+    use super::*;
+
+    #[test]
+    fn can_render_text_using_str() {
+        let list = List::new(vec![
+            ListItem::new("Item 1"),
+            ListItem::new("Item 2"),
+            ListItem::new("Item 3"),
+        ]);
+        let mut buf = Buffer::empty(Rect::new(0, 0, 6, 3));
+        Widget::render(list, buf.area, &mut buf);
+        assert_buffer_eq!(buf, Buffer::with_lines(vec!["Item 1", "Item 2", "Item 3"]));
+    }
+
+    #[test]
+    fn can_render_text_using_paragraphs() {
+        let paragraphs = vec![
+            Paragraph::new("Item 1"),
+            Paragraph::new("Item 2"),
+            Paragraph::new("Item 3"),
+        ];
+        let items = paragraphs
+            .iter()
+            .map(|p| ListItem::new(p.clone()))
+            .collect::<Vec<ListItem>>();
+        let list = List::new(items);
+        let mut buf = Buffer::empty(Rect::new(0, 0, 6, 3));
+        Widget::render(list, buf.area, &mut buf);
+        assert_buffer_eq!(buf, Buffer::with_lines(vec!["Item 1", "Item 2", "Item 3"]));
+    }
+}
diff --git a/src/widgets/paragraph.rs b/src/widgets/paragraph.rs
index 3ef6006..f589c82 100644
--- a/src/widgets/paragraph.rs
+++ b/src/widgets/paragraph.rs
@@ -51,7 +51,7 @@ pub struct Paragraph<'a> {
     /// How to wrap the text
     wrap: Option<Wrap>,
     /// The text to display
-    text: Text<'a>,
+    pub text: Text<'a>,
     /// Scroll
     scroll: (u16, u16),
     /// Alignment of the text
diff --git a/tests/widgets_list.rs b/tests/widgets_list.rs
index 99d5bfb..1575b26 100644
--- a/tests/widgets_list.rs
+++ b/tests/widgets_list.rs
@@ -78,7 +78,7 @@ fn widgets_list_should_truncate_items() {
         state.select(case.selected);
         terminal
             .draw(|f| {
-                let list = List::new(case.items.clone())
+                let list = List::new(case.items)
                     .block(Block::default().borders(Borders::RIGHT))
                     .highlight_symbol(">> ");
                 f.render_stateful_widget(list, Rect::new(0, 0, 8, 2), &mut state);
```

*mostly - there was something I ran into that prevented the derived traits on the items that's probably easily fixable.

Reply:
https://github.com/joshka/ratatui/commit/e56d8eab4c10c5e24e4596af1e7189ed630a2747 works. I.e. using List with any widget or with anything that implements Into<Text> without change. It builds on top of #122 which makes `Widget.draw()` borrow rather than consume self though, which hasn't landed yet.

I haven't handled the auto-sizing stuff in that commit though, so it's hard coded when passing a widget through to the ListItem. E.g.:

```rust
    #[test]
    fn can_render_text_using_str() {
        let list = List::new(vec![
            ListItem::new("Item 1"),
            ListItem::new("Item 2"),
            ListItem::new("Item 3"),
        ]);
        let mut buf = Buffer::empty(Rect::new(0, 0, 6, 3));
        Widget::render(&list, buf.area, &mut buf);
        assert_eq!(buf, Buffer::with_lines(vec!["Item 1", "Item 2", "Item 3"]));
    }

...

    #[test]
    fn can_render_blocks() {
        let blocks = vec![
            Block::default().title("Item 1").borders(Borders::ALL),
            Block::default()
                .title("Item 2")
                .borders(Borders::ALL)
                .title_alignment(Alignment::Center),
            Block::default()
                .title("Item 3")
                .borders(Borders::ALL)
                .title_alignment(Alignment::Right),
        ];
        let items = blocks
            .iter()
            .map(|b| WidgetWrapper::new(b.clone(), 3, 15))
            .map(|w| ListItem::new(w))
            .collect::<Vec<ListItem>>();
        let list = List::new(items);
        let mut buf = Buffer::empty(Rect::new(0, 0, 15, 9));
        Widget::render(&list, buf.area, &mut buf);
        assert_buffer_eq!(
            buf,
            Buffer::with_lines(vec![
                "┌Item 1───────┐",
                "│             │",
                "└─────────────┘",
                "┌───Item 2────┐",
                "│             │",
                "└─────────────┘",
                "┌───────Item 3┐",
                "│             │",
                "└─────────────┘",
            ])
        );
    }
```

Reply:
I have updated the `WidgetList` with the missing properties. so on API level we are backward compatible now.
I updated also the `List` logic to reuse the code of `WidgetList`.

There is a bug in the old `List` Logic when `Corner::BottomLeft` is set. This bug doesn't exit in the new Logic. the following video shows that the `WARNING` is overlapping with the upper edge. **Time 00:07**
[![asciicast](https://asciinema.org/a/GiVLe8csQq3IPOuO5mIhzaGrv.svg)](https://asciinema.org/a/GiVLe8csQq3IPOuO5mIhzaGrv)

Some unit tests are still failing.

for work I need the mixed widget for a small project, which is already done.
should I continue working on this MR?  AFAIK there is the following tasks on my list:

- StatefulWidget implementation / investigation
- Write some unitests
- Fix `List` unittests
- I need feedback from you regarding the generic types `WidgetListItem` .. `WidgetListItem5`. whether it is acceptable.

did I miss something?

Reply:
Answers to this in a top level comment on this PR below...

Reply:
nice thank you for the review